File: problem_26.py
# LeetCode Problem: Decode String
# Given an encoded string, return its decoded string.
# The encoding rule is: k[encoded_string], where the encoded_string inside the square brackets is repeated exactly k times.
# You may assume the input string is always valid and contains only digits, letters, and square brackets.

# Example:
# Input: s = "3[a]2[bc]"
# Output: "aaabcbc"

# Time Complexity: O(N)
# Space Complexity: O(N)

class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        
        for char in s:
            if char != "]":
                stack.append(char)
            else:
                curr = ""
                
                while stack and stack[-1] != "[":
                    curr = stack.pop() + curr
                
                stack.pop()
                
                num = ""
                while stack and stack[-1].isdigit():
                    num = stack.pop() + num
                
                stack.append(curr * int(num))
        
        return "".join(stack)

# Test the solution
sol = Solution()
s = "3[a]2[bc]"
result = sol.decodeString(s)
print(result)  # Output: "aaabcbc"

File: problem_26.java
// LeetCode Problem: Decode String
// Given an encoded string, return its decoded string.
// The encoding rule is: k[encoded_string], where the encoded_string inside the square brackets is repeated exactly k times.
// You may assume the input string is always valid and contains only digits, letters, and square brackets.

// Example:
// Input: s = "3[a]2[bc]"
// Output: "aaabcbc"

// Time Complexity: O(N)
// Space Complexity: O(N)

import java.util.Stack;

public class problem_26 {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();
        
        for (char ch : s.toCharArray()) {
            if (ch != ']') {
                stack.push(String.valueOf(ch));
            } else {
                StringBuilder curr = new StringBuilder();
                
                while (!stack.isEmpty() && !stack.peek().equals("[")) {
                    curr.insert(0, stack.pop());
                }
                
                stack.pop(); 
                
                StringBuilder num = new StringBuilder();
                while (!stack.isEmpty() && Character.isDigit(stack.peek().charAt(0))) {
                    num.insert(0, stack.pop());
                }
                
                int count = Integer.parseInt(num.toString());
                String repeated = curr.toString().repeat(count);
                
                stack.push(repeated);
            }
        }
        
        StringBuilder result = new StringBuilder();
        while (!stack.isEmpty()) {
            result.insert(0, stack.pop());
        }
        
        return result.toString();
    }

    public static void main(String[] args) {
        problem_26 sol = new problem_26();
        String s = "3[a]2[bc]";
        String result = sol.decodeString(s);
        System.out.println(result);  // Output: "aaabcbc"
    }
}

